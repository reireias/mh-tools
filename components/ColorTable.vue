<template>
  <table class="color-table">
    <caption>
      太字 = 合計スキルレベルが同じ範囲での最大期待値
    </caption>
    <tr>
      <td></td>
      <td v-for="(label, index) in xlabels" :key="index">{{ label }}</td>
    </tr>
    <tr v-for="(row, index) in data" :key="index">
      <td>{{ ylabels[index] }}</td>
      <td
        v-for="(value, index2) in row"
        :key="index2"
        :style="getStyle(index, index2, value)"
      >
        {{ value }}
      </td>
    </tr>
  </table>
</template>

<script>
export default {
  props: {
    xlabels: {
      type: Array,
      required: true
    },
    ylabels: {
      type: Array,
      required: true
    },
    data: {
      type: Array,
      required: true
    }
  },
  data() {
    return {
      colors: ['#BBDEFB', '#F0F4C3', '#F8BBD0']
    }
  },
  computed: {
    maxValues() {
      const values = []
      for (let i = 0; i < this.xlabels.length + this.ylabels.length - 1; i++) {
        values.push([])
      }
      this.data.forEach((row, index) => {
        row.forEach((value, index2) => {
          values[index + index2].push(value)
        })
      })
      return values.map((subValues) => {
        return Math.max(...subValues)
      })
    }
  },
  methods: {
    getStyle(row, col, value) {
      let style = `background-color: ${this.colors[(row + col) % 3]};`
      if (this.maxValues[row + col] === Number(value)) {
        style += 'font-weight: bold;'
      }
      return style
    }
  }
}
</script>

<style>
.color-table {
  border: solid 1px;
  border-collapse: collapse;
}
.color-table td {
  border: solid 1px;
  text-align: center;
  padding: 10px;
}
@media screen and (min-width: 0px) and (max-device-width: 640px) {
  .color-table td {
    font-size: 10px;
    padding: 2px;
  }
}
</style>
