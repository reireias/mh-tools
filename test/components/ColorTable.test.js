import { mount } from '@vue/test-utils'
import ColorTable from '@/components/ColorTable'

describe('ColorTable', () => {
  const props = {
    xlabels: ['A', 'B', 'C'],
    ylabels: ['X', 'Y', 'Z'],
    data: [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ],
  }

  it('should be vue instance', () => {
    const wrapper = mount(ColorTable, {
      propsData: props,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('maxValues', () => {
    it('should be max value array', () => {
      const wrapper = mount(ColorTable, {
        propsData: props,
      })
      expect(wrapper.vm.maxValues).toStrictEqual([1, 4, 7, 8, 9])
    })
  })
})
