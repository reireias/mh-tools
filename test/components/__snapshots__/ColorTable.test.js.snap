// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorTable should be vue instance 1`] = `
<table
  class="color-table"
>
  <caption>
    
    太字 = 合計スキルレベルが同じ範囲での最大期待値
  
  </caption>
   
  <tr>
    <td />
     
    <td>
      A
    </td>
    <td>
      B
    </td>
    <td>
      C
    </td>
  </tr>
   
  <tr>
    <td>
      X
    </td>
     
    <td
      style="background-color: rgb(187, 222, 251); font-weight: bold;"
    >
      
      1
    
    </td>
    <td
      style="background-color: rgb(240, 244, 195);"
    >
      
      2
    
    </td>
    <td
      style="background-color: rgb(248, 187, 208);"
    >
      
      3
    
    </td>
  </tr>
  <tr>
    <td>
      Y
    </td>
     
    <td
      style="background-color: rgb(240, 244, 195); font-weight: bold;"
    >
      
      4
    
    </td>
    <td
      style="background-color: rgb(248, 187, 208);"
    >
      
      5
    
    </td>
    <td
      style="background-color: rgb(187, 222, 251);"
    >
      
      6
    
    </td>
  </tr>
  <tr>
    <td>
      Z
    </td>
     
    <td
      style="background-color: rgb(248, 187, 208); font-weight: bold;"
    >
      
      7
    
    </td>
    <td
      style="background-color: rgb(187, 222, 251); font-weight: bold;"
    >
      
      8
    
    </td>
    <td
      style="background-color: rgb(240, 244, 195); font-weight: bold;"
    >
      
      9
    
    </td>
  </tr>
</table>
`;
